### PR TITLE
Use pcic-react-components selectors

### LINKED ISSUE #339: Dataset selection dropdown do not handle missing HadGEM2-ES data as expected
To reproduce:

1. Select "HadGEM2-ES" as the active model.
2. Select "historical, rcp85" as the active emissions scenario. 
3. Observe that all variables are selectable; none are greyed out.
4. Select "tasmax" as the active variable.
5. No data will be loaded; various messages about how data cannot be loaded appear.

There actually isn't any tasmax data for this emissions scenario and model. It shouldn't be a selectable option on the dropdown, but it is.

```sql
SELECT DISTINCT
  variable_aliases.variable_standard_name
FROM 
  ce_meta.data_file_variables, 
  ce_meta.variable_aliases, 
  ce_meta.data_files, 
  ce_meta.runs, 
  ce_meta.models, 
  ce_meta.emissions, 
  ce_meta.ensemble_data_file_variables, 
  ce_meta.ensembles
WHERE 
  data_file_variables.data_file_id = data_files.data_file_id AND
  variable_aliases.variable_alias_id = data_file_variables.variable_alias_id AND
  runs.model_id = models.model_id AND
  runs.run_id = data_files.run_id AND
  runs.emission_id = emissions.emission_id AND
  ensemble_data_file_variables.data_file_variable_id = data_file_variables.data_file_variable_id AND
  ensembles.ensemble_id = ensemble_data_file_variables.ensemble_id AND
  emissions.emission_short_name LIKE '%,%rcp85' AND 
  models.model_short_name = 'HadGEM2-ES' AND 
  ensembles.ensemble_name = 'ce_files'
ORDER BY variable_standard_name;
```

| variable_standard_name |
|------------------------|
| "altcddETCCDI"         |
| "altcsdiETCCDI"        |
| "altcwdETCCDI"         |
| "altwsdiETCCDI"        |
| "cddETCCDI"            |
| "cooling_degree_days"  |
| "csdiETCCDI"           |
| "cwdETCCDI"            |
| "dtrETCCDI"            |
| "fdETCCDI"             |
| "freezing_degree_days" |
| "growing_degree_days"  |
| "gslETCCDI"            |
| "heating_degree_days"  |
| "idETCCDI"             |
| "prcptotETCCDI"        |
| "r10mmETCCDI"          |
| "r1mmETCCDI"           |
| "r20mmETCCDI"          |
| "r95pETCCDI"           |
| "r99pETCCDI"           |
| "rp20pr"               |
| "rp20tasmax"           |
| "rp20tasmin"           |
| "rp50pr"               |
| "rp5pr"                |
| "rp5tasmax"            |
| "rp5tasmin"            |
| "rx1dayETCCDI"         |
| "rx5dayETCCDI"         |
| "sdiiETCCDI"           |
| "suETCCDI"             |
| "tn10pETCCDI"          |
| "tn90pETCCDI"          |
| "tnnETCCDI"            |
| "tnxETCCDI"            |
| "trETCCDI"             |
| "tx10pETCCDI"          |
| "tx90pETCCDI"          |
| "txnETCCDI"            |
| "txxETCCDI"            |
| "wsdiETCCDI"           |

Hopefully @rod-glover 's work on #332 will fix this issue as well.

### PR DESCRIPTION
**Pull request by rod-glover**
At long last we substitute the pcic-react-components selectors for the current ones. 

See https://github.com/pacificclimate/climate-explorer-frontend/issues/332. Some particularly gnarly selectors have not been replaced in this PR.
Fixes https://github.com/pacificclimate/climate-explorer-frontend/issues/325
Fixes #339 
Fixes https://github.com/pacificclimate/climate-explorer-frontend/issues/334 (incidentally)

This gives us the more featureful selectors and, flowing from the simplification of selectors' client code (app controllers, etc.), substantial additional improvements in the CE code, including:

- logically simpler code
- remove need for AppMixin (React recommends strongly against mixins)
- replace outdated React.createClass usage with standard ES6 class declaration
- use React recommended best practice for managing asyn data fetching (metadata)
- use HOCs to factor out some common, and somewhat complicated, code shared between app controllers

Which is all like, Yay!

What's not so Yay is that there are quite a few changes. The real change count (discounting package-lock.json) is about +762, -1035, but that's still quite a bit. Here's a guide to the substantive changes:

- DRY up and simplify app controllers, using the following. Despite the fact that there is a slightly higher line count, these components are now logically much simpler. Given the comprehensive changes to them, it's best not to look at the diffs, just at the changed code.

  - add HOC `withAsyncData`: Extracts the React best practice for managing async data fetching. It is applied to all app controllers now and can be applied in at least a couple of other places (in another PR). This may be moved to pcic-react-components.

  - Replace model, emissions scenario, and variable selectors with components from pcic-react-components. These components internally handle constraining (disabling) options based on selections in a cascade of selectors. This has enabled a lot of confusing code in app controllers to be removed.

  - Extract remaining common code in app controllers to app-controllers/common.js.

- Replace TimeOfYearSelector (in graphs and tables only; there's another similar one in MapSettings that needs replacement, but later...) with a new component (same name) in pcic-react-components.

- Make some components more robust to undefined values for props, which can happen now while selectors are doing cascading defaults.

- Fix a couple of subtle bugs related to mishandling asynchronous events.

- Add a lot of comments to new and existing code.

### CODE REVIEW

**Comment by corviday** [comment]
Haven't had time for a full review yet, but a couple quick notes from playing with the demo:

- This appears to fix #339 as well, hooray! I was hoping that would happen.

- This PR makes the `menuGroup` option in the variable configuration file obsolete; perhaps remove the option from `public/variable-options.yaml` ?

- I am in the process of replacing the remaining time series datasets with multi-year versions, so I don't expect that to be a useful category to sort the menu by, going forward.

- I like having the variables in help match the order of the variables in the dropdown to make it easier for a user user at the help to find the variable in the menu, or a user looking at the menu to find the variable in the help. I don't think we want to change the order of the variables in the Help, since they're grouped by similarity, but maybe it would make sense to have a variable index at the beginning of the variable help that gives the variables in the same order as the menu, to make it easier for someone who is looking up a variable after seeing it in the menu find it in the help?

**Comment by rod-glover** [comment]
@corviday , thanks for your prompt feedback. In response:

- `menuGroup`: Or possibly we retain this and (I think in a future PR, this one's big enough) use them to order/group the menu options. See issue https://github.com/pacificclimate/climate-explorer-frontend/issues/340.

- Do I understand rightly that all datasets will be MYM going forward?

- Re. Help: Good idea. Alternatively, we use `menuGroup`, or a variation on it that I have in mind, to order/group the menu items so that they parallel the help. See issue https://github.com/pacificclimate/climate-explorer-frontend/issues/340.

**Comment by corviday** [comment]
All datasets in the ensembles used by PCEX should be MYM as of sometime next week, I hope.

File: src/components/app-controllers/PrecipAppController/PrecipAppController.js [hunk 1]
```diff
@@ -160,31 +178,79 @@ export default createReactClass({
...
+  };
+
+  // TODO: https://github.com/pacificclimate/climate-explorer-frontend/issues/122
+  handleChangeArea = setState(this, 'area');
+  handleChangeModel = setState(this, 'model');
+  handleChangeScenario = setState(this, 'scenario');
+  handleChangeVariable = setState(this, 'variable');
+
+  render() {
+    console.log('### PrecipAppController')
```
Comment (corviday) [inline src/components/app-controllers/PrecipAppController/PrecipAppController.js]: Not sure if you meant to leave these in.

File: src/components/app-controllers/SingleAppController/SingleAppController.js [hunk 2]
```diff
@@ -1,80 +1,91 @@
...
+  replaceInvalidScenario = findScenarioIncluding('rcp85');
+  replaceInvalidVariable = findVariableMatching(opt => !opt.isDisabled);
+
+  representativeValue = (...args) => representativeValue(...args)(this.props);
+  constraintsFor = (...args) => constraintsFor(...args)(this.props);
+  filterMetaBy = (...args) =>
+    filterMetaBy(...args)(this.props)(this.props.meta);
+
+  render() {
+    console.log('### SingleAppControllerDisplay')
```
Comment (corviday) [inline src/components/app-controllers/SingleAppController/SingleAppController.js]: another debug output, unsure if you meant to leave it in.

File: src/components/app-controllers/SingleAppController/SingleAppController.js [hunk 3]
```diff
@@ -152,29 +167,63 @@ export default createReactClass({
...
+  };
+
+  // TODO: https://github.com/pacificclimate/climate-explorer-frontend/issues/122
+  handleChangeArea = setState(this, 'area');
+  handleChangeModel = setState(this, 'model');
+  handleChangeScenario = setState(this, 'scenario');
+  handleChangeVariable = setState(this, 'variable');
+
+  render() {
+    console.log('### SingleAppController')
```
Comment (corviday) [inline src/components/app-controllers/SingleAppController/SingleAppController.js]: intentional?

File: src/components/graphs/AnnualCycleGraph/AnnualCycleGraph.js [hunk 4]
```diff
@@ -137,6 +135,12 @@ export default class AnnualCycleGraph extends React.Component {
 
   // Data fetching
 
+  // TODO: Factor this out (used in other components)
```
Comment (corviday) [inline src/components/graphs/AnnualCycleGraph/AnnualCycleGraph.js]: I think this is done?

File: src/components/graphs/AnnualCycleGraph/AnnualCycleGraph.js [hunk 5]
```diff
@@ -75,48 +80,41 @@ export default class AnnualCycleGraph extends React.Component {
       instance: this.instance,  // for debugging
       prevMeta: null,
       prevArea: null,
-      prevDataSpec: null,
-      dataSpec: null,
+      prevDataspec: null,
+      dataspec: undefined,
```
Comment (corviday) [inline src/components/graphs/AnnualCycleGraph/AnnualCycleGraph.js]: You've made `dataspec` all lowercase in this graph, but it remains camelcase in other graph code. I'm worried that's going to trip me up sometime in the future.

Maybe add an issue that we're transitioning toward all-lowercase spelling of `dataspec`, to remind future programmers both 1) to make that change when they happen to be monkeying around with graph files, and 2) as a reminder that this is in process (ie, they should expect to encounter varying capitalization across files)?

Alternately, if it seems quick and easy enough, you could switch all the graphs now.

Comment (rod-glover) [reply src/components/graphs/AnnualCycleGraph/AnnualCycleGraph.js]: Good catch; an artefact of introducing the new selector.

**Pull request merged**